docs/proposals/affinity.md

```diff
@@ -0,0 +1,341 @@
+Repeat Step 1b except replace the last line with "foreach node W of {N} not having label [key=H.TopologyKey, value=any K such that M[K]>0], Y[W]++"
+// Step 4: Symmetric soft, plus treat forward direction of hard affinity as a soft
+foreach node A of {N}
+	foreach pod B that is bound to A
+		increment Y[A] by the number of B's SoftAffinityExpressions, SoftAntiAffinityExpressions, and HardAffinityExpressions that are satisfied if Z runs on A but are not satisfied if Z does not run on A
+// We're done. {N} contains all of the nodes that satisfy the affinity/anti-affinity rules, and Y is
+// a map whose keys are the elements of {N} and whose values are how "good" of a choice N is for Z with
+// respect to the explicit and implicit affinity/anti-affinity rules (larger number is better).
+```
+
+## Implementation plan
+
+1. Implement the API resource
+1. Implement respect for AffinitySelector.HardAffinityExpressions and AffinitySelector.HardAntiAffinityExpressions as a new scheduler predicate function
+1. Implement respect for AffinitySelector.SoftAffinityExpressions and AffinitySelector.SoftAntiAffinityExpressions as a new scheduler priority function
+1. Implement a workaround for the issue described at the end of the Affinity section of the Examples section (can't schedule first pod)
+
+## Backward compatiblity
+
+Old versions of the scheduler will ignore `AffinitySelector`.
+
+## Future work and non-work
+
+The hard anti-affinity API we have described only allows rules of the form "allow
+no more than one pod matching PodSelector PS to run in this topology domain." The
+common use case specifies an empty PodSelector and a topology domain of "node," thus
+"do not allow this pod to share a node with any other pods." One can imagine that
+for larger topology domains, one might want to associate a number with this rule,
+for example "do not allow this pod to share a rack with more than three other
+pods (in total, or from the same service as the pod)." We could allow this to be
+specified by adding an integer `Limit` to `AffinityTerm` just for the
+`HardAntiAffinityExpressions` case. However, this flexibility complicates the
+system and we do not intend to implement it.
+
+A malicious or misconfigured pod can cause serious problems via HardAntiAffinity. For example,
+imagine that a new cluster is created, and the first user creates a pod P with hard anti-affinity
+against all other pods in the cluster. This will block any other pods from scheduling in the cluster. One
+approach to addressing this vulnerability is for an admission control component to reject pods with problematic
+hard anti-affinity settings. Another is to implement a priority/preemption scheme, which we
+eventually need to do anyway. (Note that these two approaches are not mutually exclusive.)
+In such a scheme, some pods have higher scheduling priority than others; pods with higher
+scheduling priority than P will be able to evict P in order to schedule.
```

Comment:
Related to this is the "pay for what you use" problem, which potentially helps to solve the "malicious or misconfigured" problem you mention.   More specifically, if one wanted to charge the owner of a pod for the resources that their pod consumes, surely one should charge them for the whole node if they express such a hard anti-affinity constraint as per the above example, which precludes any other pod from using the node?  In which case, an alternative formulation for them would be not to use such an anti-affinity constraint, but rather to specify minimum resource requirements corresponding to an entire node (which would count against their quota and for which they would be appropriately charged - it would also preclude other pods from co-habitating on the node, as there would be no unallocated resources remaining).  Of course they might also need to specify a node type constraint, to prevent being scheduled onto larger nodes along with undesirable neighbors.  I'm rambling now, so I'll shut up :)